### PR TITLE
Start and clean up workers from the local scheduler.

### PR DESCRIPTION
**Pull request by stephanie-wang**
Workers can be started from either the local scheduler or the Python script, as before. Workers send their process IDs to the local scheduler upon initial connection, so that the local scheduler can clean up worker child processes.

### CODE REVIEW

File: python/ray/services.py [hunk 1]
```diff
@@ -89,18 +90,24 @@ def kill_process(p):
   if RUN_PHOTON_PROFILER or RUN_PLASMA_MANAGER_PROFILER or RUN_PLASMA_STORE_PROFILER:
     os.kill(p.pid, signal.SIGINT) # Give process signal to write profiler data.
     time.sleep(0.1) # Wait for profiling data to be written.
-  p.kill()
-  # Sleeping for 0 should yield the core and allow the killed process to process
-  # its pending signals.
-  time.sleep(0)
+
+  # Allow the process one second to exit gracefully.
+  p.terminate()
```
Comment (robertnishihara) [inline python/ray/services.py]: Why reverse the order of `kill` and `terminate`? Doesn't it make sense to start with `kill`?

File: python/ray/worker.py [hunk 2]
```diff
@@ -1523,6 +1523,7 @@ def process_task(task): # wrapping these lines in a function should cause the lo
                                         "function_name": function_name})
 
   check_main_thread()
+  worker.photon_client.register_pid()
```
Comment (robertnishihara) [inline python/ray/worker.py]: I think it is more appropriate for this to be part of `photon_connect`. I needed something similar for the actor PR (to pass an "actor ID" from the worker to the local scheduler), so we should probably use the same mechanism.

I did it like this https://github.com/ray-project/ray/pull/242/commits/5b2275077060b4097288d3c637cad0f300e994a2#diff-2a906576526ef8fb8a2de97b6e0bc3ceR7

What do you think about doing it that way? (Your PR will be ready first)

NOTE: the local scheduler should never terminate a driver, so if we do it in `photon_connect`, we will need the option of passing `-1` or something to indicate that it is a driver. (but this will be useful anyway because the local scheduler really ought to know which clients are drivers (for computing resource capacity).

Comment (robertnishihara) [inline python/ray/worker.py]: Also, the purpose of this should be documented (e.g., "We pass the process ID to the local scheduler so that the local scheduler can terminate this worker if it wants to.")

File: src/photon/photon_scheduler.c [hunk 3]
```diff
@@ -23,8 +24,148 @@
...
 UT_icd workers_icd = {sizeof(local_scheduler_client *), NULL, NULL, NULL};
 
+UT_icd pid_t_icd = {sizeof(pid_t), NULL, NULL, NULL};
+
 UT_icd byte_icd = {sizeof(uint8_t), NULL, NULL, NULL};
 
+void kill_process(pid_t pid) {
+  /* Kill the worker violently. */
+  kill(pid, SIGKILL);
+  waitpid(pid, NULL, 0);
```
Comment (robertnishihara) [inline src/photon/photon_scheduler.c]: I'm a little nervous that in pathological situations this could cause the local scheduler to hang forever.

Even during normal usage, do we really want to local scheduler to block here?

File: src/photon/photon.h [hunk 4]
```diff
@@ -50,7 +44,7 @@ typedef struct scheduling_algorithm_state scheduling_algorithm_state;
  *  scheduler. */
 typedef struct {
   /** The script to use when starting a new worker. */
-  char *start_worker_command;
+  const char **start_worker_command;
```
Comment (robertnishihara) [inline src/photon/photon.h]: Why split the command into different strings? Is this so you can use `fork` instead of `popen`?

File: src/photon/photon_scheduler.c [hunk 5]
```diff
@@ -23,8 +24,148 @@
...
+ * Start a new worker as a child process.
+ *
+ * @param state The state of the local scheduler.
+ * @return Void.
+ */
+void start_new_worker(local_scheduler_state *state) {
+  /* We can't start a worker if we don't have the path to the worker script. */
+  CHECK(state->config.start_worker_command != NULL);
+  /* Launch the process to create the worker. */
+  pid_t pid = fork();
```
Comment (robertnishihara) [inline src/photon/photon_scheduler.c]: Why use `fork` instead of `popen`? With `fork` the new worker process will have some leftover state from the local scheduler, right?

File: python/ray/services.py (hunk 1 above)
Comment (stephanie-wang) [inline python/ray/services.py]: `kill` is stronger than `terminate`. I'm doing `terminate` first to give the local scheduler a chance to catch the signal and clean up state. Previously it wasn't really an issue because it just needed to free memory that would get cleaned up on exit anyway, but now there are worker processes to clean up.

File: python/ray/worker.py (hunk 2 above)
Comment (stephanie-wang) [inline python/ray/worker.py]: Oh yeah, makes sense to do it as part of `photon_connect`.

The local scheduler won't ever terminate the driver. It only terminates child processes. If you look at the handler for `REGISTER_PID` messages, it matches up worker clients with child processes according to pid.

File: src/photon/photon_scheduler.c (hunk 3 above)
Comment (stephanie-wang) [inline src/photon/photon_scheduler.c]: Yeah, I'm not sure about this either...ideally we would mark the worker as dead, and then clean up the rest of the state after receiving some signal that it died. I can try to look into that.

File: src/photon/photon_scheduler.c (hunk 5 above)
Comment (stephanie-wang) [inline src/photon/photon_scheduler.c]: It calls `execvp` soon after, which if successful, completely replaces the process image.

`popen` didn't seem necessary in this case, since we're already communicating with the process through sockets. Also, I needed the PID to be returned to the local scheduler so that it could later figure out if a worker process was a child or not.

File: src/photon/photon.h (hunk 4 above)
Comment (stephanie-wang) [inline src/photon/photon.h]: It's for `execvp`.

File: src/photon/photon_scheduler.c [hunk 6]
```diff
@@ -345,6 +509,8 @@ void new_client_connection(event_loop *loop,
   local_scheduler_client *worker = malloc(sizeof(local_scheduler_client));
   worker->sock = new_socket;
   worker->task_in_progress = NULL;
+  worker->pid = 0;
```
Comment (robertnishihara) [inline src/photon/photon_scheduler.c]: Can a normal PID equal 0?

File: src/photon/photon.h [hunk 7]
```diff
@@ -84,6 +81,10 @@ typedef struct {
    *  no task is running on the worker, this will be NULL. This is used to
    *  update the task table. */
   task *task_in_progress;
+  /** The process ID of the client. */
```
Comment (robertnishihara) [inline src/photon/photon.h]: Maybe document that if this is 0, that means that the client has not connected to the local scheduler yet.

File: src/photon/photon_scheduler.c (hunk 6 above)
Comment (stephanie-wang) [inline src/photon/photon_scheduler.c]: Not any worker PID. PID 0 is reserved by the kernel.

File: src/photon/photon_scheduler.c [hunk 8]
```diff
@@ -500,6 +669,13 @@ int main(int argc, char *argv[]) {
   if (!node_ip_address) {
     LOG_FATAL("please specify the node IP address with -h switch");
   }
+  int num_workers = 0;
+  if (num_workers_str) {
+    num_workers = strtol(num_workers_str, NULL, 10);
```
Comment (robertnishihara) [inline src/photon/photon_scheduler.c]: We should check that `strtol` succeeded and that `num_workers_str` was not malformed

File: src/photon/test/photon_tests.c [hunk 9]
```diff
@@ -476,12 +515,77 @@ TEST task_multi_dependency_test(void) {
   PASS();
 }
 
+TEST start_kill_workers_test(void) {
+  /* Start some workers. */
+  int num_workers = 4;
```
Comment (atumanov) [inline src/photon/test/photon_tests.c]: we should add a separate test (a stress test if you will) that tries to do this with 100 workers. I've tried it with 100 workers just now, and it wasn't working for me.

File: src/photon/test/photon_tests.c [hunk 10]
```diff
@@ -476,12 +515,77 @@ TEST task_multi_dependency_test(void) {
...
+  process_message(photon->photon_state->loop, worker->sock, worker, 0);
+  ASSERT_EQ(utarray_len(photon->photon_state->child_pids), 0);
+  ASSERT_EQ(utarray_len(photon->photon_state->workers), num_workers);
+
+  /* Clean up. */
+  close(new_worker_fd);
+  destroy_photon_mock(photon);
+  PASS();
+}
+
```
Comment (atumanov) [inline src/photon/test/photon_tests.c]: would be great to add a test that dynamically changes the number of workers : e.g.,
start 20 workers
repeat {
  kill 2 workers
  add 1 worker
} until num_workers > 0
This dynamic test will make sure we are not leaking worker state at runtime and that we're accounting for workers correctly (with asserts before/after each _worker operation )

**Comment by atumanov** [comment]
ok, we've discussed with Stephanie that we should do a more graceful shutdown of workers when we try to kill or evict them. It's best to send the worker a SIGTERM signal and implement a SIGTERM handler in the worker to allow the worker to close all of its sockets. In the meantime, the local scheduler schedules a timer to send a followup SIGKILL to the evicted worker after 1ms. The reason for this is that, as things are right now, the worker ends up in an inconsistent state and attempts to communicate on sockets that are still open, fails and generates core files while failing. This is reproduceable on a Mac with num_workers >= 10 . Example core dump:

```
(lldb) bt
* thread #1: tid = 0x0000, 0x00007fff9b1b5f06 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff9b1b5f06 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff931a04ec libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff8bda26df libsystem_c.dylib`abort + 129
    frame #3: 0x0000000105d91281 libphoton.so`photon_connect(photon_socket="/tmp/photon_socket_1115438165") + 321 at photon_client.c:14
    frame #4: 0x0000000105d8e4ee libphoton.so`PyPhotonClient_init(self=0x00000001002ac3d8, args=0x00000001034e3bd0, kwds=0x0000000000000000) + 78 at photon_extension.c:24
    frame #5: 0x0000000100062465 libpython2.7.dylib`type_call + 229
    frame #6: 0x000000010000c761 libpython2.7.dylib`PyObject_Call + 97
    frame #7: 0x00000001000a35a5 libpython2.7.dylib`PyEval_EvalFrameEx + 4405
    frame #8: 0x00000001000ac7dc libpython2.7.dylib`PyEval_EvalCodeEx + 2092
    frame #9: 0x00000001000ac988 libpython2.7.dylib`fast_function + 296
    frame #10: 0x00000001000a338a libpython2.7.dylib`PyEval_EvalFrameEx + 3866
    frame #11: 0x00000001000ac7dc libpython2.7.dylib`PyEval_EvalCodeEx + 2092
    frame #12: 0x00000001000ac856 libpython2.7.dylib`PyEval_EvalCode + 54
    frame #13: 0x00000001000cb3d4 libpython2.7.dylib`PyRun_FileExFlags + 164
    frame #14: 0x00000001000cc709 libpython2.7.dylib`PyRun_SimpleFileExFlags + 409
    frame #15: 0x00000001000e01ca libpython2.7.dylib`Py_Main + 2938
    frame #16: 0x0000000100000f14 python`start + 52
```
code where this worker is failing:
```
  pid_t my_pid = getpid();
  int success = write_message(result->conn, REGISTER_PID, sizeof(my_pid),
                              (uint8_t *) &my_pid);
  CHECKM(success == 0, "Unable to register worker with local scheduler");
  return result;
}
```

Another example core dump:
```
(lldb) bt
* thread #1: tid = 0x0000, 0x00007fff9b1b5f06 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff9b1b5f06 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff931a04ec libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff8bda26df libsystem_c.dylib`abort + 129
    frame #3: 0x0000000105d8f739 libphoton.so`photon_get_task(conn=0x000000010184fc20) + 297 at photon_client.c:58
    frame #4: 0x0000000105d8c5b2 libphoton.so`PyPhotonClient_get_task(self=0x00000001002ac3d8) + 34 at photon_extension.c:49
    frame #5: 0x00000001000a2f66 libpython2.7.dylib`PyEval_EvalFrameEx + 2806
    frame #6: 0x00000001000ac7dc libpython2.7.dylib`PyEval_EvalCodeEx + 2092
    frame #7: 0x00000001000ac988 libpython2.7.dylib`fast_function + 296
    frame #8: 0x00000001000a338a libpython2.7.dylib`PyEval_EvalFrameEx + 3866
    frame #9: 0x00000001000ac7dc libpython2.7.dylib`PyEval_EvalCodeEx + 2092
    frame #10: 0x00000001000ac856 libpython2.7.dylib`PyEval_EvalCode + 54
    frame #11: 0x00000001000cb3d4 libpython2.7.dylib`PyRun_FileExFlags + 164
    frame #12: 0x00000001000cc709 libpython2.7.dylib`PyRun_SimpleFileExFlags + 409
    frame #13: 0x00000001000e01ca libpython2.7.dylib`Py_Main + 2938
    frame #14: 0x0000000100000f14 python`start + 52
```
Here, the worker is trying to send the GET_TASK message to photon .

Takeaway, we should let the worker clean up its socket connections and gracefully self-terminate.

File: src/photon/photon_scheduler.c [hunk 11]
```diff
@@ -53,6 +56,184 @@ void print_resource_info(const local_scheduler_state *state,
...
+    }
+  }
+
+  /* Clean up the client socket after killing the worker so that the worker
+   * can't receive the SIGPIPE before exiting. */
+  close(worker->sock);
+
+  /* Clean up the task in progress. */
+  if (worker->task_in_progress) {
+    /* TODO(swang): Update the task table to mark the task as lost. */
```
Comment (atumanov) [inline src/photon/photon_scheduler.c]: So, this will stall all the downstream tasks that depend on any objects produced by this task_in_progress. We should really have a mechanism for re-scheduling the task in flight. One possibility is to 
just update the task table entry for this task with status = TASK_STATUS_KILLED and subscribe to this task status in the global scheduler. The GS handler would then re-schedule the task.
Without this, we are likely to see hangs.

Comment (atumanov) [inline src/photon/photon_scheduler.c]: also, this task will never report done back to photon, so we need to do resource accounting for this task. Otherwise we will leak dynamic resources.

Comment (atumanov) [inline src/photon/photon_scheduler.c]: basically, you need this code in this if block:
```
      task_spec *spec = task_task_spec(worker->task_in_progress);
      /* Return dynamic resources back for the task in progress. */
      for (int i = 0; i < MAX_RESOURCE_INDEX; i++) {
        state->dynamic_resources[i] += task_spec_get_required_resource(spec, i);
        /* Sanity-check resource vector boundary conditions. */
        CHECK(state->dynamic_resources[i] <= state->static_resources[i]);
      }
```

Comment (stephanie-wang) [inline src/photon/photon_scheduler.c]: Yes, this PR is a little incomplete right now, since workers are only killed during cleanup, when reporting task death doesn't matter anymore. I'm hoping to implement reporting task death properly in a separate PR for task preemption.

Comment (atumanov) [inline src/photon/photon_scheduler.c]: the resource accounting bit should still be addressed. I'm happy to do it in a separate PR, along with photon_tests updates that would expose and test this condition.

File: src/photon/photon_scheduler.c [hunk 12]
```diff
@@ -53,6 +56,184 @@ void print_resource_info(const local_scheduler_state *state,
...
+ * Kill a worker, if it is a child process, and clean up all of its associated
+ * state.
+ *
+ * @param worker A pointer to the worker we want to kill.
+ * @param wait A bool representing whether we should wait for the worker's
+ *        process to exit. If the worker is not a child process, this flag is
+ *        ignored.
+ * @return Void.
+ */
+void kill_worker(local_scheduler_client *worker, bool wait) {
```
Comment (robertnishihara) [inline src/photon/photon_scheduler.c]: When we kill a worker, we probably need to let the scheduling algorithm do some clean up as well (in particular, the scheduling algorithm needs to remove the worker struct from its list of available workers, otherwise it will try to assign a task to that worker).
